configure: add option to enable/disable libnuma

jstebbins opened the pull request:
It is not supported or not useful in some Linux environments, so allow
disabling it.

**Test on:**

- [x] Windows 10+  (via MinGW)
- [x] macOS 10.13+
- [x] Ubuntu Linux

Should only affect linux, but wouldn't hurt to do a smoke test on other platforms.

jstebbins committed:
configure: add option to enable/disable libnuma

It is not supported or not useful in some Linux environments, so allow
disabling it.

bradleysepos commented on make/configure.py:
```diff
@@ -1238,6 +1238,10 @@ def createCLI( cross = None ):
     grp.add_argument( '--enable-x265', dest="enable_x265", default=True, action='store_true', help=(( 'enable %s' %h ) if h != argparse.SUPPRESS else h) )
     grp.add_argument( '--disable-x265', dest="enable_x265", action='store_false', help=(( 'disable %s' %h ) if h != argparse.SUPPRESS else h) )
 
+    h = IfHost( 'x265 NUMA support', '*-*-linux', none=argparse.SUPPRESS ).value
```
If NUMA is Linux and Windows compatible:

`h = 'x265 NUMA support' if host.match( '*-*-linux*' ) or (cross is not None and 'mingw' in cross) else argparse.SUPPRESS`

If NUMA is Linux-only:

`h = 'x265 NUMA support' if host.match( '*-*-linux*' ) and cross is None else argparse.SUPPRESS`

bradleysepos commented on make/configure.py:
```diff
@@ -1238,6 +1238,10 @@ def createCLI( cross = None ):
     grp.add_argument( '--disable-x265', dest="enable_x265", action='store_false', help=(( 'disable %s' %h ) if h != argparse.SUPPRESS else h) )
 
+    h = IfHost( 'x265 NUMA support', '*-*-linux', none=argparse.SUPPRESS ).value
+    grp.add_argument( '--enable-numa', dest="enable_numa", default=True, action='store_true', help=(( 'enable %s' %h ) if h != argparse.SUPPRESS else h) )
```
Default for Linux and Windows:

`default=(host.match( '*-*-linux*' ) or (cross is not None and 'mingw' in cross))`

Default for Linux only:

`default=(host.match( '*-*-linux*' ) and cross is None)`

jstebbins replied:
I've taken the more difficult approach of fixing build vs host notation.  So of coarse much more testing will now be needed on all platforms :tongue:

bradleysepos replied:
Might still want to make the default platform-specific here.

jstebbins replied:
Fixed

bradleysepos reviewed (changes requested):
Couple inline comments.

IfHost() is not sufficient to detect whether we are cross-compiling. A match on Linux machine could actually be building for Windows. This is partly why I updated `configure.py` to make the `cross` variable available early.

Thoughts on ripping out all uses of IfHost() to avoid future confusion?

Could also avoid these lengthy conditionals by sorting it out at the top of the function and setting a single variable to represent the system, similar to how we have `build.system` later on in the main section.

jstebbins committed:
configure: add PkgConfigProbe and ChkLib actions

PkgConfigProbe checks for the existence of a lib using pkg-config.
ChkLib tries to verify a lib with PlkConfigProbe, then tries with
LDProbe.

Use ChkLib to verify existence of libnuma when configured and fail if
not found

jstebbins committed:
configure: Fix host vs build specs

Fixes the notation internally in configure.py.  Does not yet fix the
notation written to GNUMakefile or m4 files since that requires
modification of multiple files.  Save this for the next pass.

jstebbins commented:
It appears Travis does not have libnuma installed?  My configure check seems to indicate that.

bradleysepos commented:
Nice, I wanted to do this in my last PR but didn't want to jump ahead too far. Looks good on initial read.

The "is" operator compares identity, not value. It's a stricter comparison. See PEP-8: https://www.python.org/dev/peps/pep-0008/#programming-recommendations

> Comparisons to singletons like None should always be done with is or is not, never the equality operators.
>
> Also, beware of writing if x when you really mean if x is not None -- e.g. when testing whether a variable or argument that defaults to None was set to some other value. The other value might have a type (such as a container) that could be false in a boolean context!

jstebbins committed:
configure: initialize host_tuple with cross

Must be done before createCLI

jstebbins committed:
configure: use stricter 'is not None'

jstebbins committed:
configure: move init of HostTuple after parsing cross

jstebbins committed:
configure: make sure numa is only enabled on linux

jstebbins commented on make/configure.py:
```diff
@@ -579,13 +693,13 @@ def _action( self ):
-        elif host.match( '*-*-freebsd.*' ):
-            self.mode['i386']   = 'i386-portsbuild-freebsd%s' % (host.release)
-            self.mode['amd64'] = 'amd64-portsbuild-freebsd%s' % (host.release)
+        elif host_tuple.match( '*-*-freebsd.*' ):
```
@bradleysepos does this look right to you?  i.e. `*-*-freebsd.*`.  Seems like the '.' shouldn't be there.  I didn't change this, just noticed it while reviewing my own code for typos.

bradleysepos replied:
Looks like a typo. Feel free to fix.

jstebbins committed:
configure: move sanitation of options together

and fix some typos

jstebbins committed:
configure: fix typo

bradleysepos reviewed (commented):
Left the comments behind...

jstebbins committed:
configure: fix comments

jstebbins committed:
configure: make separate LDTest and PkgConfigTest functions

Using LDProbe and PkgConfigProbe Actions within ChkLib Action resulted
in somewhat ugly log output.  Moving the work into separate Test
functions allows it to be reused without using the Actions

jstebbins committed:
configure: Add comments to make lib checks more readable

With all the quoted code fragments, it is difficult to see indentation
that separates osx, mingw, and linux specific tests.  The comments make
this more clear

bradleysepos commented:
Tested all platforms.

jstebbins commented:
> Tested all platforms.

Fantastic! Thanks.
